### PR TITLE
Fix some iframe edge cases

### LINKED ISSUE #13648: 16.5 with better support of iframe has some side effects
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

When transitioning from a state having an iframe inside a component to a state without that said iframe, we have an error at `setOffsets(node, offsets)` because `node.ownerDocument.defaultView` being `null`

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**

Working on it

**What is the expected behavior?**

Should not break

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

Affected since this 16.5.0, coming from this PR exactly https://github.com/facebook/react/pull/12037/files#diff-26d90cba6ee597ef475fa80dcf76ae1d

`getOffsets` as a check on `ownerDocument.default` fallbacks to `window`, but `setOffsets` does not.

Is there a reason why?

### PR DESCRIPTION
**Pull request by JSteunou**
Should fix #13648 by being as protective in `setOffsets` than in `getOffsets`

### CODE REVIEW

**Comment by gaearon** [comment]
Can you verify this fixes it? Also it would really be helpful to add a case to this fixture:

https://github.com/facebook/react/tree/master/fixtures/dom/src/components/fixtures/selection-events

File: packages/react-dom/src/client/ReactDOMSelection.js [hunk 1]
```diff
@@ -151,9 +151,12 @@ export function getModernOffsetsFromPoints(
  * @param {object} offsets
  */
 export function setOffsets(node, offsets) {
-  const doc = node.ownerDocument || document;
-  const win = doc ? doc.defaultView : window;
-  const selection = win.getSelection();
+  const {ownerDocument} = node;
```
Comment (acusti) [inline packages/react-dom/src/client/ReactDOMSelection.js]: Thanks for tracking this down! This is definitely an oversight of #12037 that we didn’t address.

For context, the reason that `getOffsets` uses a different method to get a relative `window` and `document` object than `setOffsets` is that `getOffsets` only needs to use the relative `window` (which has `getSelections`), whereas `setOffsets` needs to use both the `window` object (for `getSelection`, again), and the `document` object (for `createRange`). So this should really be a hybrid of the two approaches, i.e.:
```js
  const doc = node.ownerDocument || document;
  const win = (doc && doc.defaultView) || window;
  ...
```
Where it still guards against a missing `ownerDocument` by falling back to the global and also guards against the case  of having an `ownerDocument` that’s  missing a `defaultView`.

Comment (JSteunou) [reply packages/react-dom/src/client/ReactDOMSelection.js]: I understand, I changed the logic a bit by over-protecting this path

Update incoming

**Comment by JSteunou** [comment]
and yes the actual PR fix the issue we have

**Comment by gaearon** [comment]
@acusti Can you sign off on the final version please? If it looks good to you I'll get it in.

**Comment by gaearon** [comment]
Thanks.

**Comment by JSteunou** [comment]
Thank you both!

**Comment by einarq** [comment]
Was this tested in Edge? I saw a bug related to pasting stuff in a TinyMCE editor when using 16.5.1, upgrading to 16.5.2 solved it. But it is still failing in Edge. The issue might be related to something entirely different of course, but it goes away if I downgrade to 16.4.2.
The error coming from TinyMCE is "Unable to get property 'moveToBookmark' of undefined or null reference", not sure if that helps at all.

**Comment by acusti** [comment]
@einarq Can you post a link to a repo or even just a code sample that makes it possible to reproduce the error on Edge? I’m most interested in how you are including and rendering the TinyMCE editor, but anything that will help me to reproduce it. If the error is coming from TinyMCE itself, I’m not sure how it’s related to the changes between 16.4.2 and 16.5.x, but I would be happy to take a look to try to figure it out.

**Comment by einarq** [comment]
I’ll see if I can create a repro and provide some more information next week, but I was just mainly wondering if the fix for the issue was tested in Edge (assuming there was a repro for the bug that was fixed).

**Comment by einarq** [comment]
I have traced the problem to the change in setOffsets between 16.4.2 and
16.5.0.

As previously mentioned we have a TinyMCE editor in every row of a list,
which supports paste to add more rows. When pasting just 5 rows, setOffsets
does not get called at all, but when pasting for instance 100 rows,
setOffsets gets called and fails with "Object expected". Perhaps some type
of timing issue?

A can try to create an example but I suspect it would be difficult due to
the complexity of the scenario with TinyMCE being used etc.

Could we first check if this code even works in Edge? Do you have a repro
case for this bug that was fixed, and if so perhaps that could be tested
with Edge?

Thanks,

Einar

**Comment by einarq** [comment]
It fails due to win.getSelection being undefined.
I see that in 16.4.2 this function had an early return statement for when
getSelection is undefined, could we add that back in?

Changing setOffsets to this makes it work again for me:

  const doc = node.ownerDocument || document;
  const win = doc ? doc.defaultView : window;

  if (!win.getSelection) {
    return;
  }

**Comment by acusti** [comment]
@einarq Thanks for tracking that down! If you want to make a PR that adds an early return `if (!win.getSelection)` to the top of `setOffsets` with the explanation of what it addresses, I imagine that would be easy to get merged. I’m puzzled as to why it’s still necessary, considering that support for `window.getSelection` was added in IE9, and:
```js
const win = (doc && doc.defaultView) || window;
```
will ensure that even if `doc.defaultView` is not a proper `Window` object, we’ll still get the `window` global, which must have the `getSelection` method on it. Only thing I could imagine is that `doc.defaultView` is present but is some kind of limited version of a `Window` object that doesn’t have the full APIs, maybe for cross-domain security reasons?

But again, having an early return there is harmless, so I’m fine with it, even though I don’t understand why it’s needed.

**Comment by einarq** [comment]
Done. Uncharted territory for me (until now), so hope I did it correctly.
https://github.com/facebook/react/pull/14095

Thanks in advance for any help in getting it merged :)

Einar

**Pull request merged**